pkg/api/parser/protobuf/protobuf.go

```diff
@@ -12,14 +13,24 @@ import (
 // ParseRequest is responsible for populating the write request from the
 // data in the request in protobuf format.
 func ParseRequest(r *http.Request, wr *prompb.WriteRequest) error {
-	reqBuf, err := ioutil.ReadAll(r.Body)
+	b := bufPool.Get().(*bytes.Buffer)
+	b.Reset()
+
+	_, err := b.ReadFrom(r.Body)
 	if err != nil {
 		return fmt.Errorf("request body read error: %w", err)
 	}
 
-	if err = proto.Unmarshal(reqBuf, wr); err != nil {
+	if err = proto.Unmarshal(b.Bytes(), wr); err != nil {
 		return fmt.Errorf("protobuf unmarshal error: %w", err)
 	}
 
-	return nil
+	bufPool.Put(b)
```

Comment:
I think this should be in `defer` (line 17) as if we error out in line `21` or `25`, we do not put in the pool and this gets GCed.

Reply:
good catch